crawling/groupBus.py

```diff
@@ -0,0 +1,92 @@
+from scipy.spatial import KDTree
+import json
+import math
+import polars as pl
+
+def haversine_distance(lat1, lon1, lat2, lon2):
+    R = 6371000  # Earth radius in meters
+    
+    phi1 = math.radians(lat1)
+    phi2 = math.radians(lat2)
+    delta_phi = math.radians(lat2 - lat1)
+    delta_lambda = math.radians(lon2 - lon1)
+    
+    a = math.sin(delta_phi/2)**2 + math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda/2)**2
+    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
+    
+    return R * c
+
+def calculate_bearing(lat1, lon1, lat2, lon2):
+    phi1 = math.radians(lat1)
+    phi2 = math.radians(lat2)
+    delta_lambda = math.radians(lon2 - lon1)
+    
+    y = math.sin(delta_lambda) * math.cos(phi2)
+    x = math.cos(phi1) * math.sin(phi2) - math.sin(phi1) * math.cos(phi2) * math.cos(delta_lambda)
+    theta = math.atan2(y, x)
+    
+    return (math.degrees(theta) + 360) % 360
+
+def group_bus_stops(bus_stops, max_distance=50, bearing_threshold=35):
+    tree = KDTree(bus_stops.select("lat", "lng"))
+    groups = pl.DataFrame(schema={"id":str, "lat":pl.Float64, "lng":pl.Float64, "name_en":str, "name_zh":str, "bus_group_id":pl.Int32})
+    visited = set()
+    group_id = 1
+
+    print(len(bus_stops))
```

Comment:
It is observed that from start of the script to this print line, it took roughly 3 minutes to run on GitHub actions. I would suspect the KDTree is taking most of the time